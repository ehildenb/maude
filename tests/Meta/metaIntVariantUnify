#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/metaIntVariantUnify.maude -no-banner -no-advise \
  > metaIntVariantUnify.out 2>&1

diff $srcdir/metaIntVariantUnify.expected metaIntVariantUnify.out > /dev/null 2>&1
