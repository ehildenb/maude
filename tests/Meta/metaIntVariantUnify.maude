set show timing off .
set show advisories off .

***
*** Test variant unification in the meta-interpreter.
***

load metaInterpreter

fmod XOR is
  sort XOR .

  sort Elem .
  ops c1 c2 c3 c4 : -> Elem .

  subsort Elem < XOR .
  op _+_ : XOR XOR -> XOR [ctor assoc comm] .
  op 0 : -> XOR .
  op a : -> XOR .

  vars X Y : XOR .

  eq Y + 0 = Y [variant] .
  eq X + X = 0 [variant] .
  eq X + X + Y = Y [variant] .
endfm

mod VARIANT-UNIFY-TEST is
  pr META-INTERPRETER .
  sort MsgList .
  subsort Msg < MsgList .
  op me : -> Oid .
  op User : -> Cid .
  op problem:(_,_) : UnificationProblem TermList -> Attribute .
  op soln:_ : Nat -> Attribute .
  op result : MsgList -> Attribute .
  op _,_ : MsgList MsgList -> MsgList [assoc id: nil] .
  op nil : -> MsgList .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  var N : Nat .
  var V : Term .
  var TL : TermList .
  var S : Substitution .
  var Q : Qid .
  var B : Bool .
  var P : Parent .
  var R : RewriteCount .
  var ML : MsgList .
  var UP : UnificationProblem .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('XOR, true)) .

  rl < X : User | AS, problem:(UP, TL) > insertedModule(X, Y) =>
     < X : User | AS, problem:(UP, TL), soln: 1, result(nil) > getVariantUnifier(Y, X, 'XOR, UP, TL, '#, 0) .

  rl < X : User | AS, problem:(UP, TL), soln: N, result(ML) > gotVariantUnifier(X, Y, R, S, Q) =>
     < X : User | AS,  problem:(UP, TL), soln: (N + 1), result(ML, gotVariantUnifier(X, Y, R, S, Q)) >
     getVariantUnifier(Y, X, 'XOR, UP, TL, '#, N) .
endm

erew in VARIANT-UNIFY-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'c1.Elem] =? '_+_['Y:XOR, 'c2.Elem], empty) > 
createInterpreter(interpreterManager, me, none) .

variant unify in XOR : X:XOR + c1 =? Y:XOR + c2 .

erew in VARIANT-UNIFY-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'c1.Elem] =? '_+_['Y:XOR, 'c2.Elem], '_+_['X:XOR, 'c1.Elem]) > 
createInterpreter(interpreterManager, me, none) .

variant unify in XOR : X:XOR + c1 =? Y:XOR + c2 such that X:XOR + c1 irreducible .

mod DISJOINT-VARIANT-UNIFY-TEST is
  pr META-INTERPRETER .
  sort MsgList .
  subsort Msg < MsgList .
  op me : -> Oid .
  op User : -> Cid .
  op problem:(_,_) : UnificationProblem TermList -> Attribute .
  op soln:_ : Nat -> Attribute .
  op result : MsgList -> Attribute .
  op _,_ : MsgList MsgList -> MsgList [assoc id: nil] .
  op nil : -> MsgList .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  var N : Nat .
  var V : Term .
  var TL : TermList .
  vars S S' : Substitution .
  var Q : Qid .
  var B : Bool .
  var P : Parent .
  var R : RewriteCount .
  var ML : MsgList .
  var UP : UnificationProblem .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('XOR, true)) .

  rl < X : User | AS, problem:(UP, TL) > insertedModule(X, Y) =>
     < X : User | AS, problem:(UP, TL), soln: 1, result(nil) > getDisjointVariantUnifier(Y, X, 'XOR, UP, TL, '#, 0) .

  rl < X : User | AS, problem:(UP, TL), soln: N, result(ML) > gotDisjointVariantUnifier(X, Y, R, S, S', Q) =>
     < X : User | AS,  problem:(UP, TL), soln: (N + 1), result(ML, gotDisjointVariantUnifier(X, Y, R, S, S', Q)) >
     getDisjointVariantUnifier(Y, X, 'XOR, UP, TL, '#, N) .
endm

erew in DISJOINT-VARIANT-UNIFY-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'c1.Elem] =? '_+_['Y:XOR, 'c2.Elem], empty) > 
createInterpreter(interpreterManager, me, none) .

erew in DISJOINT-VARIANT-UNIFY-TEST : <> 
< me : User | problem: ('_+_['X:XOR, 'c1.Elem] =? '_+_['Y:XOR, 'c2.Elem], '_+_['X:XOR, 'c1.Elem]) > 
createInterpreter(interpreterManager, me, none) .
