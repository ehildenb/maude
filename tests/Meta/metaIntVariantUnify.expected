==========================================
erewrite in VARIANT-UNIFY-TEST : <> < me : User | problem:('_+_['X:XOR,
    'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],empty) > createInterpreter(
    interpreterManager, me, none) .
saw '_+_['X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem]
rewrites: 43
result Configuration: <> < me : User | soln: 9,result(gotVariantUnifier(me,
    interpreter(0), 6, 
  'X:XOR <- '_+_['c2.Elem,'@1:XOR] ; 
  'Y:XOR <- '_+_['c1.Elem,'@1:XOR], '@),gotVariantUnifier(me, interpreter(0),
    0, 
  'X:XOR <- 'c2.Elem ; 
  'Y:XOR <- 'c1.Elem, '@),gotVariantUnifier(me, interpreter(0), 18, 
  'X:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR] ; 
  'Y:XOR <- '%1:XOR, '%),gotVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '%1:XOR ; 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR], '%),gotVariantUnifier(me,
    interpreter(0), 0, 
  'X:XOR <- '0.XOR ; 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem], '@),gotVariantUnifier(me, interpreter(0),
    0, 
  'X:XOR <- 'c1.Elem ; 
  'Y:XOR <- 'c2.Elem, '@),gotVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem,'c2.Elem] ; 
  'Y:XOR <- '0.XOR, '@),gotVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem,'@1:XOR] ; 
  'Y:XOR <- '_+_['c2.Elem,'@1:XOR], '@)),problem:('_+_['X:XOR,'c1.Elem] =?
    '_+_['Y:XOR,'c2.Elem],empty) > noSuchResult(me, interpreter(0), 0, true)
==========================================
variant unify in XOR : X + c1 =? Y + c2  .

Unifier #1
rewrites: 6
X --> c2 + %1:XOR
Y --> c1 + %1:XOR

Unifier #2
rewrites: 6
X --> c2
Y --> c1

Unifier #3
rewrites: 24
X --> c1 + c2 + #1:XOR
Y --> #1:XOR

Unifier #4
rewrites: 24
X --> #1:XOR
Y --> c1 + c2 + #1:XOR

Unifier #5
rewrites: 24
X --> 0
Y --> c1 + c2

Unifier #6
rewrites: 24
X --> c1
Y --> c2

Unifier #7
rewrites: 24
X --> c1 + c2
Y --> 0

Unifier #8
rewrites: 24
X --> c1 + %1:XOR
Y --> c2 + %1:XOR

No more unifiers.
rewrites: 24
==========================================
erewrite in VARIANT-UNIFY-TEST : <> < me : User | problem:('_+_['X:XOR,
    'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],'_+_['X:XOR,'c1.Elem]) >
    createInterpreter(interpreterManager, me, none) .
saw '_+_['X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem]
rewrites: 12
result Configuration: <> < me : User | soln: 4,result(gotVariantUnifier(me,
    interpreter(0), 3, 
  'X:XOR <- '_+_['c2.Elem,'@1:XOR] ; 
  'Y:XOR <- '_+_['c1.Elem,'@1:XOR], '@),gotVariantUnifier(me, interpreter(0),
    0, 
  'X:XOR <- 'c2.Elem ; 
  'Y:XOR <- 'c1.Elem, '@),gotVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '%1:XOR ; 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR], '%)),problem:('_+_['X:XOR,
    'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],'_+_['X:XOR,'c1.Elem]) > noSuchResult(
    me, interpreter(0), 0, true)
==========================================
variant unify in XOR : X + c1 =? Y + c2  such that X + c1 irreducible .

Unifier #1
rewrites: 3
X --> c2 + %1:XOR
Y --> c1 + %1:XOR

Unifier #2
rewrites: 3
X --> c2
Y --> c1

Unifier #3
rewrites: 3
X --> #1:XOR
Y --> c1 + c2 + #1:XOR

No more unifiers.
rewrites: 3
==========================================
erewrite in DISJOINT-VARIANT-UNIFY-TEST : <> < me : User | problem:('_+_[
    'X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],empty) > createInterpreter(
    interpreterManager, me, none) .
saw '_+_['X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem]
rewrites: 43
result Configuration: <> < me : User | soln: 9,result(
    gotDisjointVariantUnifier(me, interpreter(0), 6, 
  'X:XOR <- '_+_['c2.Elem,'@1:XOR], 
  'Y:XOR <- '_+_['c1.Elem,'@1:XOR], '@),gotDisjointVariantUnifier(me,
    interpreter(0), 0, 
  'X:XOR <- 'c2.Elem, 
  'Y:XOR <- 'c1.Elem, '@),gotDisjointVariantUnifier(me, interpreter(0), 18, 
  'X:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR], 
  'Y:XOR <- '%1:XOR, '%),gotDisjointVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '%1:XOR, 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR], '%),gotDisjointVariantUnifier(me,
    interpreter(0), 0, 
  'X:XOR <- '0.XOR, 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem], '@),gotDisjointVariantUnifier(me,
    interpreter(0), 0, 
  'X:XOR <- 'c1.Elem, 
  'Y:XOR <- 'c2.Elem, '@),gotDisjointVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem,'c2.Elem], 
  'Y:XOR <- '0.XOR, '@),gotDisjointVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '_+_['c1.Elem,'@1:XOR], 
  'Y:XOR <- '_+_['c2.Elem,'@1:XOR], '@)),problem:('_+_['X:XOR,'c1.Elem] =?
    '_+_['Y:XOR,'c2.Elem],empty) > noSuchResult(me, interpreter(0), 0, true)
==========================================
erewrite in DISJOINT-VARIANT-UNIFY-TEST : <> < me : User | problem:('_+_[
    'X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],'_+_['X:XOR,'c1.Elem]) >
    createInterpreter(interpreterManager, me, none) .
saw '_+_['X:XOR,'c1.Elem] =? '_+_['Y:XOR,'c2.Elem]
rewrites: 12
result Configuration: <> < me : User | soln: 4,result(
    gotDisjointVariantUnifier(me, interpreter(0), 3, 
  'X:XOR <- '_+_['c2.Elem,'@1:XOR], 
  'Y:XOR <- '_+_['c1.Elem,'@1:XOR], '@),gotDisjointVariantUnifier(me,
    interpreter(0), 0, 
  'X:XOR <- 'c2.Elem, 
  'Y:XOR <- 'c1.Elem, '@),gotDisjointVariantUnifier(me, interpreter(0), 0, 
  'X:XOR <- '%1:XOR, 
  'Y:XOR <- '_+_['c1.Elem,'c2.Elem,'%1:XOR], '%)),problem:('_+_['X:XOR,
    'c1.Elem] =? '_+_['Y:XOR,'c2.Elem],'_+_['X:XOR,'c1.Elem]) > noSuchResult(
    me, interpreter(0), 0, true)
Bye.
